### PR TITLE
KEYCLOAK-5098 : Spring boot2

### PR DESCRIPTION
**Pull request by sebastienblanc**
This PR adds a new adapter for Spring Boot 2

### CODE REVIEW

**Comment by ryandawsonuk** [comment]
Looking good. 

Something I noticed is the tests under:

https://github.com/sebastienblanc/keycloak/tree/140a2f0ed4580c8ad8ba3c9a8b4042aea2f618cd/adapters/oidc/spring-boot/src/test/java/org/keycloak/adapters/springboot/client

They are equally applicable to spring-boot2 and could be duplicated under the adapters/oidc/spring-boot2 path. Sounds strange to duplicate but they come from lower down in adapters/oidc/spring-security so it would be testing that configuration from the keycloak-spring-security-adapter is correctly applied. There may be better ways to handle but can't immediately think of any.

**Comment by sebastienblanc** [comment]
@ryandawsonuk Good catch, I think we can just duplicate them, ;et's keep it easy and in the long term SB 1.x will be deprecated.

**Comment by sebastienblanc** [comment]
So I just added a important commit : the upgrade to 2.0.0.RELEASE , it seems all okay for now \o/ , running some final tests and we should be good to go.

**Comment by stianst** [comment]
Does this replace https://github.com/keycloak/keycloak/pull/4325? Can we close that PR?

**Comment by Salaboy** [comment]
+1

**Comment by tblakers** [comment]
Will this be compatible with Webflux (reactive Spring boot) ?

**Comment by jagin** [comment]
When we can except this PR to be merged? Please :) It would be nice to use Spring Boot 2 with Keycloak.

**Comment by iamsharife** [comment]
Yes please! +1

File: adapters/oidc/spring-boot-adapter-core/pom.xml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        <artifactId>spring-boot-configuration-processor</artifactId>
+        <optional>true</optional>
+        <version>${spring-boot.version}</version>
+     </dependency>
+  </dependencies>
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
```
Comment (pedroigor) [inline adapters/oidc/spring-boot-adapter-core/pom.xml]: IIRC, source and target are already defined in```keycloak-parent``` pom.xml. Recently I had some issues with this configuration in sub-modules and better is keep settings defined in ```keycloak-parent```.

Or is there any reason for this definition here ?

File: adapters/oidc/spring-boot2/pom.xml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-configuration-processor</artifactId>
+      <optional>true</optional>
+      <version>${spring-boot.version}</version>
+    </dependency>
+  </dependencies>
+  <build>
```
Comment (pedroigor) [inline adapters/oidc/spring-boot2/pom.xml]: Same here, inherit source/target from ```keycloak-parent```.

**Review by pedroigor — CHANGES_REQUESTED** [review]

**Comment by pedroigor** [comment]
Beside those review comments and test failures in CI. LGTM.

**Review by pedroigor — APPROVED** [review]

**Comment by iamsharife** [comment]
Approved?! - Can this be merged ...

**Comment by codependent** [comment]
We are using RedhatSOO 7.1.3.GA, will the adapter be compatible with this server?

**Comment by kossmoboleat** [comment]
I've read that beta 1 didn't include these changes. When will beta 2 or a rc be released?

**Comment by SahanDll** [comment]
Anyone have working sample code for keycloak 4.0 with spring boot 2.0
Thank you

**Comment by Vigneshsekar** [comment]
Please let us know when we can expect the next release (Beta 2). We are waiting to use Keycloak with Spring Boot 2.

**Comment by stianst** [comment]
Beta 2 should be out next week

**Comment by mithandir** [comment]
I updated to Beta2 in our application but the error seems to be not fixed? Do I need to configure something additionaly?

It ClassNotFoundException but the Source is different each time I start the application. Sometimes its EmbeddedServletContainerCustomizer sometime Tomcat, Undertow etc.

Exception: 

`java.lang.IllegalStateException: Failed to introspect Class [org.keycloak.adapters.springboot.KeycloakAutoConfiguration] from ClassLoader [sun.misc.Launcher$AppClassLoader@18b4aac2]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:659) ~[spring-core-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:556) ~[spring-core-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:541) ~[spring-core-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.util.ReflectionUtils.getUniqueDeclaredMethods(ReflectionUtils.java:599) ~[spring-core-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getTypeForFactoryMethod(AbstractAutowireCapableBeanFactory.java:724) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:665) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:633) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1489) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:420) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:390) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:511) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:503) ~[spring-beans-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.getBeansOfType(AbstractApplicationContext.java:1198) ~[spring-context-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	at org.springframework.boot.SpringApplication.getExitCodeFromMappedException(SpringApplication.java:889) [spring-boot-2.0.1.RELEASE.jar:2.0.1.RELEASE]
	at org.springframework.boot.SpringApplication.getExitCodeFromException(SpringApplication.java:875) [spring-boot-2.0.1.RELEASE.jar:2.0.1.RELEASE]
	at org.springframework.boot.SpringApplication.handleExitCode(SpringApplication.java:861) [spring-boot-2.0.1.RELEASE.jar:2.0.1.RELEASE]
	at org.springframework.boot.SpringApplication.handleRunFailure(SpringApplication.java:810) [spring-boot-2.0.1.RELEASE.jar:2.0.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:338) [spring-boot-2.0.1.RELEASE.jar:2.0.1.RELEASE]
	at ch.sbb.esta.openshift.gracefullshutdown.GracefulshutdownSpringApplication.run(GracefulshutdownSpringApplication.java:21) [springboot-graceful-shutdown-2.0.jar:na]
	at ch.sbb.nets.userprofile.service.NetsUserProfileServiceApp.main(NetsUserProfileServiceApp.java:19) [classes/:na]
Caused by: java.lang.NoClassDefFoundError: org/springframework/boot/context/embedded/EmbeddedServletContainerCustomizer
	at java.lang.Class.getDeclaredMethods0(Native Method) ~[na:1.8.0_144]
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701) ~[na:1.8.0_144]
	at java.lang.Class.getDeclaredMethods(Class.java:1975) ~[na:1.8.0_144]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:641) ~[spring-core-5.0.5.RELEASE.jar:5.0.5.RELEASE]
	... 19 common frames omitted

Caused by: java.lang.ClassNotFoundException: org.springframework.boot.context.embedded.EmbeddedServletContainerCustomizer
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[na:1.8.0_144]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_144]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335) ~[na:1.8.0_144]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_144]
	... 23 common frames omitted
`

**Comment by PhilippHeuer** [comment]
Make sure you include the spring boot starter and don't use the adapter directly - i just had the same error as you when i included the adapter instead of the spring boot starter dependency.

```
compile group: 'org.springframework.boot', name: 'spring-boot-starter-security'
compile group: 'org.keycloak', name: 'keycloak-spring-boot-2-starter', version: '4.0.0.Beta2'
```

**Comment by mithandir** [comment]
Thank you. I did use `keycloak-spring-boot-starter` instead of `keycloak-spring-boot-2-starter`.  Now its working fine.

**Pull request merged**